---
title: "ISS Data Analysis"
output: html_notebook
---

## Load libraries

```{r}
library(Seurat)
library(Matrix)
library(ggplot2)
library(tidyverse)
library(dplyr)
library(gprofiler2)
library(sctransform)
library(magrittr)
```
## Load the data

```{r}

ISS_matrix <- readMM("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/matrix.mtx")

```


```{r}
features <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/features.tsv", sep="\t", header=FALSE, stringsAsFactors=FALSE)

barcodes <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/barcodes.tsv", header=FALSE, stringsAsFactors=FALSE)

metadata <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/Fig3e-j_Xenium.csv", stringsAsFactors=FALSE, sep=";", header=TRUE, row.names=1)
```

Check if dimensions match

```{r}
nrow(ISS_matrix) == nrow(features)
ncol(ISS_matrix) == nrow(barcodes)
```

Assign barcodes and genes as names for the matrix

```{r}
rownames(ISS_matrix) <- features$V2  # gene symbols as row names
colnames(ISS_matrix) <- barcodes$V1  # barcodes as column names
colnames(features) <- c("ensembl","gene","gene_expression")
```

# 1) QC 

## Filter both the matrix and the feature list to exclude probes

First, filter the features list

```{r}
features <- filter(features, gene_expression == "Gene Expression")
```

Filter matrix based on the filtered features df

```{r}
genes_list <- features$gene

ISS_matrix_filtered <- ISS_matrix[genes_list, ]
```

Create a separate column for cell barcodes to keep track of them; remove unlabeled cells.

```{r}
metadata$barcodes <- rownames(metadata)
metadata <- metadata[metadata$Cluster != "Unlabeled", ]
```

Filter and keep cells in metadata and matrix that are present in both

```{r}
same_barcodes <- intersect(metadata$barcodes, colnames(ISS_matrix_filtered))
ISS_matrix_filtered <- ISS_matrix_filtered[, same_barcodes]
metadata <- metadata[metadata$barcodes %in% same_barcodes, ]
```

## Subset the data to only include 100 cells from each cell type

Select cells we're gonna keep (100 from every cell type)

```{r}
sampled_cells <- metadata %>%
  group_by(Cluster) %>%
  slice_sample(n = 100, replace = FALSE) %>% 
  pull(barcodes)

```

```{r}
ISS_matrix_filtered <- ISS_matrix_filtered[, colnames(ISS_matrix_filtered) %in% sampled_cells]
table(metadata$Cluster[metadata$barcodes %in% colnames(ISS_matrix_filtered)])
```


# 2) Normalization 

## Method 1: log2(x+1) normalization

```{r}
norm_ISS_matrix <- log2(ISS_matrix_filtered+1)
```

## Method 2: SCTransform

```{r}
ISS_matrix_filtered <- as(ISS_matrix_filtered, "CsparseMatrix")
```

```{r}
SCT_ISS_matrix <- sctransform::vst(ISS_matrix_filtered)$y
```

# 4) PCA analysis

## For log normalized counts

```{r}
norm_ISS_se <- CreateSeuratObject(counts = norm_ISS_matrix, 
                                 project = "XeniumProject",
                                 assay = "Xenium",
                                 meta.data = metadata)
```

Since I performed normalization outside the Seurat object, this next step is necessary for PCA. The problem is that when you create the Seurat object, it will only add the data to the RNA counts. ScaleData() uses the data from the 'data' slot, which would normally contain the normalized data (if you have performed the normalization on the se). 

This will assign the normalized data to the 'data' slot as well.

```{r}
norm_ISS_se[["Xenium"]]$data <- norm_ISS_matrix 
```

```{r}
head(norm_ISS_se[[]])
```

PCA:

```{r}
norm_ISS_se <- FindVariableFeatures(norm_ISS_se) 
norm_ISS_se <- ScaleData(norm_ISS_se)
norm_ISS_se <- RunPCA(norm_ISS_se, verbose = FALSE)
```

## SCTransformed data

```{r}
SCT_ISS_se <- CreateSeuratObject(counts = SCT_ISS_matrix, 
                                 project = "XeniumProject",
                                 assay = "Xenium",
                                 meta.data = metadata)
```

```{r}
SCT_ISS_se[["Xenium"]]$data <- SCT_ISS_matrix
```


```{r}
head(SCT_ISS_se[[]])
```

```{r}
DefaultAssay(SCT_ISS_se) <- "Xenium"
```


```{r}
SCT_ISS_se <- FindVariableFeatures(SCT_ISS_se) 
SCT_ISS_se <- ScaleData(SCT_ISS_se)
SCT_ISS_se <- RunPCA(SCT_ISS_se, verbose = FALSE)
```

# 5) Visualize PCA

## Log normalized data

```{r}
# Get the PCA data
PCA_norm_ISS <- Embeddings(norm_ISS_se, reduction = "pca")
PCA_norm_ISS  <- as.data.frame(PCA_norm_ISS )
PCA_norm_ISS$barcodes <- rownames(PCA_norm_ISS ) 

# Add metadata
metadata_PCA_norm <- norm_ISS_se@meta.data
PCA_norm_ISS  <- cbind(PCA_norm_ISS, metadata_PCA_norm)
# Rename one of the barcodes column (added them to the metadata before to keep track but now it's in the way)
colnames(PCA_norm_ISS)[ncol(PCA_norm_ISS)] <- "cell_id"
```

```{r}
colnames(PCA_norm_ISS)
```

Colored by sequencing depth:

```{r}
ggplot(PCA_norm_ISS, aes(x = PC_1, y = PC_2, color = nCount_Xenium)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Sequencing Depth", title = "Log Normalized ISS Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_norm.png",
  plot = last_plot()
  )
```

Colored by cell type:

```{r}
ggplot(PCA_norm_ISS, aes(x = PC_1, y = PC_2, color = Cluster)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Cell Type", title = "Log Normalized ISS Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_norm_cellty.png",
  plot = last_plot()
  )
```

## SCTransformed data

```{r}
# Get the PCA data
PCA_SCT_ISS <- Embeddings(SCT_ISS_se, reduction = "pca")
PCA_SCT_ISS  <- as.data.frame(PCA_SCT_ISS)
PCA_SCT_ISS$barcodes <- rownames(PCA_SCT_ISS) 

# Add metadata
metadata_PCA_SCT <- SCT_ISS_se@meta.data
PCA_SCT_ISS  <- cbind(PCA_SCT_ISS, metadata_PCA_SCT)
# Rename one of the barcodes column (added them to the metadata before to keep track but now it's in the way)
colnames(PCA_SCT_ISS)[ncol(PCA_SCT_ISS)] <- "cell_id"
```

Colored by sequencing depth:

```{r}
ggplot(PCA_SCT_ISS, aes(x = PC_1, y = PC_2, color = nCount_Xenium)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Sequencing Depth", title = "SCTransformed ISS Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_SCT.png",
  plot = last_plot()
  )
```

Colored by cell type:

```{r}
ggplot(PCA_SCT_ISS, aes(x = PC_1, y = PC_2, color = Cluster)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Cell Type", title = "SCTransformed ISS Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_SCT_cellty.png",
  plot = last_plot()
  )
```

## Elbow plot to visualise SDs of PCs 

## Log normalized

```{r}
ElbowPlot(norm_ISS_se)
ggsave(
  "elbowplot_norm.png",
  plot = last_plot()
  )
```

```{r}
# Determine percent of variation associated with each PC
pct_norm <- norm_ISS_se[["pca"]]@stdev / sum(norm_ISS_se[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu_norm <- cumsum(pct_norm)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
PC1_norm <- which(cumu_norm > 90 & pct_norm < 5)[1]

PC1_norm

# Determine the difference between variation of PC and subsequent PC
PC_diff_norm <- sort(which((pct_norm[1:length(pct_norm) - 1] - pct_norm[2:length(pct_norm)]) > 0.1), decreasing = T)[1] + 1

# Last point where change of % of variation is more than 0.1%.
PC_diff_norm
```

## SCTransformed

```{r}
ElbowPlot(SCT_ISS_se)
ggsave(
  "elbowplot_SCT.png",
  plot = last_plot()
  )
```


```{r}

# Determine percent of variation associated with each PC
pct_SCT <- SCT_ISS_se[["pca"]]@stdev / sum(SCT_ISS_se[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu_SCT <- cumsum(pct_SCT)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
PC1_SCT <- which(cumu_SCT > 90 & pct_SCT < 5)[1]

PC1_SCT

# Determine the difference between variation of PC and subsequent PC
PC_diff_SCT <- sort(which((pct_SCT[1:length(pct_SCT) - 1] - pct_SCT[2:length(pct_SCT)]) > 0.1), decreasing = T)[1] + 1

# Last point where change of % of variation is more than 0.1%.
PC_diff_SCT
```

# 6) Run a UMAP on the PCA data - set a threshold for the PCs, e.g. 90% (based on the elbow plot)

Color UMAP by cell type - write a fn to quantify it (to see the separation of clusters) 

## LogNormalized

```{r}
norm_ISS_se <- RunUMAP(norm_ISS_se, dims = 1:17)  # Adjust based on elbow plot
```


```{r}
umap_df_norm <- as.data.frame(Embeddings(norm_ISS_se, reduction = "umap"))
```

```{r}
# Add metadata (from se to preserve consistency)
metadata_umap_norm <- norm_ISS_se@meta.data
umap_df_norm <- cbind(umap_df_norm, metadata_umap_norm)
```

Colored by sequencing depth

```{r}
ggplot(umap_df_norm, aes(x = umap_1, y = umap_2, color = nCount_Xenium)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Sequencing Depth", title = "SCTransformed Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "umap_norm_seqdepth.png",
  plot = last_plot()
  )
```

Colored by cell type:

```{r}
ggplot(umap_df_norm, aes(x = umap_1, y = umap_2, color = Cluster)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Cell Type", title = "SCTransformed Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "umap_norm_cellty.png",
  plot = last_plot()
  )
```

## SCTranformed

```{r}
SCT_ISS_se <- RunUMAP(SCT_ISS_se, dims = 1:16)  # Adjust based on elbow plot
```

```{r}
umap_df_SCT <- as.data.frame(Embeddings(SCT_ISS_se, reduction = "umap"))
```

```{r}
# Add metadata (from se to preserve consistency)
metadata_umap_SCT <- SCT_ISS_se@meta.data
umap_df_SCT <- cbind(umap_df_SCT, metadata_umap_SCT)
```

Colored by sequencing depth

```{r}
ggplot(umap_df_SCT, aes(x = umap_1, y = umap_2, color = nCount_Xenium)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Sequencing Depth", title = "SCTransformed Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "umap_SCT_seqdepth.png",
  plot = last_plot()
  )
```

Colored by cell type:

```{r}
ggplot(umap_df_SCT, aes(x = umap_1, y = umap_2, color = Cluster)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Cell Type", title = "SCTransformed Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "umap_SCT_cellty.png",
  plot = last_plot()
  )
```

## UMAP quantificiation:
    - every cell in a cluster to other cells in the clusters
    - every cell to other cell type cells in other clusters
    - plot this to see the overlaps
 
```{r}
quant <- function(umap_df) {
  # Calculating the pairwise distances, storing in dist matrix
  dist_matrix <- as.matrix(dist(umap_df[, c("umap_1", "umap_2")]))
  
  # Calculating distances (to cells in same cluster and in other clusters) for a cell
  dist_calc <- function(i) {
    cell_cluster <- umap_df$Annotation[i]
    same_cluster <- umap_df$Annotation == cell_cluster #logical vector
    
    # Storing two vectors of distances per cell in a list: 1) dist to cells in same cluster, 
    # 2) dist to cells in other clusters
    list(
      in_cluster = dist_matrix[i, same_cluster & (seq_along(same_cluster) != i)],
      other_clusters = dist_matrix[i, !same_cluster]
    )
  }
  
  # Apply the function to all cells
  results <- lapply(seq_len(nrow(umap_df)), dist_calc)
  names(results) <- rownames(umap_df) #assign correct barcodes to the resulting list
  
  return(results)
}
```

## SCTransformed

```{r}
# Apply the quantification fn 
quant_SCT <- quant(umap_df_SCT)
```

## Log normalized

```{r}
# Apply the quantification fn 
quant_norm <- quant(umap_df_norm)
```


# kNN clustering — for every cell, 5 closest neighbors

## LogNormalized

```{r}
norm_ISS_se <- FindNeighbors(
  norm_ISS_se,
  reduction = "umap",
  dims = 1:2,
  k.param = 5)
```

## SCTransformed

```{r}
SCT_ISS_se <- FindNeighbors(
  SCT_ISS_se,
  reduction = "umap",
  dims = 1:2,
  k.param = 5)
```



# Projecting normalized scRNA-seq data on to the PCA from the ISS data

## Log normalized data


I need to filter scRNA-seq to only contain the genes also present in ISS, right? Do they need to be comparable in that sense? Would it even be possible to do this without filtering of genes? The dimensions wouldn't match?

- ISS contains 313 genes
- scRNA-seq 

Find common genes? Are all the genes selected in ISS present in the scRNA-seq data?

Options:

1) normalized scRNA-seq matrix on to the PCA from ISS - filter genes based on PCA from ISS

2) normalized and PCA-transformed scRNA-seq matrix on to the PCA from ISS - find common genes


1)

```{r}
RPM_normalized_matrix_sc <- readRDS(file = "C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/RPM_normalized_matrix_sc.Rdata")
SCT_normalized_matrix_sc <- readRDS(file = "C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/SCT_normalized_matrix_sc.Rdata")
```

## Log normalized

```{r}
# Get the PCA data
ISS_PCA_matrix <- Embeddings(SCT_ISS_se, reduction = "pca")
ISS_PCA_matrix <- as(ISS_PCA_matrix,"CsparseMatrix")
```

PCA coordinates extracted in this way describe how the cells from the ISS data are arranged in the PCA space. Since I need to project my scRNA-seq data into the same PCA space, rather than the coordinates associated with the cells, I need to know which genes and how contribute to the PCA components. 

For that, I'm gonna get the PCA loadings instead (which indicate the contribution of each original feature to the principal components). And I can't even 

We are interested in seeing how the scRNA-seq data would be arranged in the PCA space from the ISS data, right??

```{r}
ISS_PCA_loadings <- Loadings(SCT_ISS_se, reduction = "pca")
ISS_PCA_loadings <- as(ISS_PCA_loadings,"CsparseMatrix")
```

```{r}
common_genes_1 <- intersect(rownames(ISS_PCA_loadings), rownames(RPM_normalized_matrix_sc))
ISS_PCA_loadings <- ISS_PCA_loadings[common_genes_1, ]
RPM_normalized_matrix_sc <- RPM_normalized_matrix_sc[common_genes_1, ]
```

```{r}
dim(RPM_normalized_matrix_sc)
dim(ISS_PCA_loadings)
```


```{r}
RPM_normalized_projected <- t(RPM_normalized_matrix_sc) %*% ISS_PCA_loadings
```

Problem is, I don't know how to actually do the projection - I tried projecting the scRNA-seq data (cells x genes) to the genes x PCs matrix. Doesn't work because of the different dimensions. Transposing doesn't help, still different dimensions - can I even use this approach if one dimension of the matrices is different?

There is also a Seurat function for integrating datasets which I haven't tried yet. 




Another question - should I use the normalized matrix or dim.red./normalized data? I understand that both are possible and valid, but which option are we more interested in?






