---
title: "scRNA-seq Data Analysis"
output: html_notebook
---

# Loading libraries
```{r load_seurat}
  library(Seurat)
  library(gprofiler2)
  library(ggplot2)
  library(dplyr)
  library(magrittr)
  library(Matrix)
  library(viridis)
  library(pheatmap)
  library(sctransform)
  library(future)
  library(igraph)
```
```{r}
ggsave <- function(..., bg = 'white') ggplot2::ggsave(..., bg = bg)
```


# Data specifics

1) matrix file - contains the nr of transcripts per gene, per cell (barcode)

2) features file - gene ID, gene name, feature type

3) barcodes - barcode sequences

4) metadata - annotations

# 1) Loading the data

```{r}
count_matrix <- readMM("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/scRNA-seq/data/matrix.mtx")

features <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/scRNA-seq/data/features.tsv", sep="\t", header=FALSE, stringsAsFactors=FALSE)

barcodes <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/scRNA-seq/data/barcodes.tsv", header=FALSE, stringsAsFactors=FALSE)
```

## Checking if the dimensions match

```{r}
nrow(count_matrix) == nrow(features)
ncol(count_matrix) == nrow(barcodes)
```

Looks good, can set names for rows and columns 

```{r}
rownames(count_matrix) <- features$V2  # gene symbols as row names
colnames(count_matrix) <- barcodes$V1  # barcodes as column names
```

## Also load metadata

```{r}
metadata <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/scRNA-seq/data/Fig2a_scFFPE-seq_UMAP.csv", stringsAsFactors=FALSE, sep=";", header=TRUE, row.names=1)

head(metadata)
```

## Creating an assay object before creating the Seurat object 

Objective to remove duplicate rownames as the Seurat object cannot be created before doing this

```{r}
count_assay <- CreateAssayObject(count_matrix)
```

# 2) Creating a Seurat object

```{r}
se <- CreateSeuratObject(counts = count_assay, meta.data = metadata)
```

```{r}
head(se[[]])
```

Remove cells with no annotation

```{r}
se <- subset(x = se, Annotation != "NA")
head(se[[]])
```

# 3) Quality Control

To understand where to set the threshold for filtering, it's a good idea to visualize the distribution of our data

## Distribution of unique genes per cell

```{r}
gene_dist_plot <- hist(log10(se$nFeature_RNA+1), breaks = 50, main = "Distribution of Unique Genes Per Cell", xlab = "Unique Genes Per Cell")

png(file = "C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/scRNA-seq/figures/dist_genes.png", width = 800, height = 500)
plot(gene_dist_plot)
dev.off()
```
Pretty low gene counts mostly? Need to filter to reduce the low-count genes

## Log transformed count depth

```{r}
count_depth_plot <- hist(log10(se$nCount_RNA+1), breaks = 50, main = "Distribution of Count Depth Per Cell", xlab = "Count Per Cell")

png(file = "C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/scRNA-seq/figures/dist_countdepth.png", width = 800, height = 500)
plot(gene_dist_plot)
dev.off()
```

# 1) Filter cells - more than 2000 reads

```{r}
# Selecting cells with more than 2000 reads
cell_set <- se$nCount_RNA > 2000
keep_cells <- colnames(se)[cell_set]
head(keep_cells)

# Using these cell names to subset the Seurat object
se_subset <- subset(se, cells = keep_cells)
head(se_subset[[]])

```

## Filter out lowly expressed genes

```{r}
# Fetching the filtered count matrix
filt_mat <- LayerData(se_subset, assay = "RNA", layer = "counts")
# head(filt.mat)

# Finding the percentage of cells expressing each gene
expression_percentage <- rowMeans(filt_mat > 0) * 100

# Selecting genes that are expressed in at least 1% of cells
genes_filt <- names(expression_percentage[expression_percentage > 1])
se_subset <- subset(se_subset, features = genes_filt)
```


# 3) Content of mitochondrial transcripts - compare exonic and mitochondrial fractions - keep the majority of the cells but remove a few outliers (with high mitochondrial counts) 

## Using pattern searching to identify and filter cells with high mitochondrial gene counts (went with this instead of biomart)

```{r}
# Using pattern searching to identify the mitochondrial genes
# Calculating percentages of mitochondrial genes per cell
se_subset[["percent_mito"]] <- PercentageFeatureSet(object = se_subset, pattern = "^MT-")

se_subset[[]]
```

Filtering out cells in which mitochondrial gene expression exceeds 15%

```{r}
selected_mito <- WhichCells(se_subset, expression = percent_mito < 15) # threshold based on Janesick paper
se_subset <- subset(se_subset, cells = selected_mito)
```

```{r}
head(se_subset[[]])
```

# 2) Normalization - log transformed RPM - normalize to median of the depth 

## Extract the filtered data from the Seurat object first

```{r}
count_matrix_filt <- LayerData(se_subset, assay = "RNA", layer = "counts")
head(count_matrix_filt)
```

## Normalization using SCTransform

Too large for normalization currently, to combat this:

1) Increased memory usage

```{r}
options(future.globals.maxSize = 1600 * 1024^2)  
```


2) Parallelization with the 'future' package - normalization spread over 4 workers

3) Apply vst or variance stabilizing transformation from sctransform 

```{r}
plan() #check current plan
```

```{r}
# Set a new plan
plan(multisession, workers = 4) #workers sounds so cute :(

# Apply vst from sctransform
normalized_data <- sctransform::vst(count_matrix_filt)$y #$y to only access the normalized matrix
```

Convert to the sparse format to save on memory usage

```{r}
SCT_normalized <- as(normalized_data, "dgCMatrix")
```

## Normalization with NormalizeData (LogNormalize):

log(count / total count per cell) * (median total count + 1)

Instead of normalizing to a fixed value of million, normalize to the median of the sequencing depth across all cells.

After, log transformation to stabilize variance. 

LogNormalize basically employs the RPM/CPM method followed by log transformation; just replaced the scale factor with median depth 

```{r}
# Median depth across all cells
median.depth <- median(se_subset$nCount_RNA)

plan(sequential)

# And then Seurat's normalization
se_RPM_norm <- NormalizeData(
  object = se_subset,
  normalization.method = "LogNormalize",
  scale.factor = median.depth,
  verbose = FALSE
)
```


# 4) Subset the data (100 of every subtype)

## SCTransform normalized data

Fetch metadata from se_subset (because if I used the original metadata, the random sampling might include cells that were already filtered out)

```{r}
metadata_se <- se_subset@meta.data
all(rownames(metadata_se) == colnames(SCT_normalized))  #see if cell barcodes match
```

Get a list of 100 cells per cell type to keep

```{r}
metadata_se$barcodes <- rownames(metadata_se)

sampled_cells <- metadata_se %>%
  group_by(Annotation) %>%
  slice_sample(n = 100, replace = FALSE) %>% 
  pull(barcodes)

#sample_cells
```

Use sampled_cells to subset 

```{r}
SCT_normalized_subset <- SCT_normalized[, colnames(SCT_normalized) %in% sampled_cells]
```

```{r}
# Create an assay object before creating a Seurat object
SCT_normalized_assay <- CreateAssayObject(SCT_normalized_subset) 
se_SCT_norm_subset <- CreateSeuratObject(SCT_normalized_assay, meta.data = metadata)

# Check if the subsetting worked
Idents(se_SCT_norm_subset) <- se_SCT_norm_subset$Annotation
table(Idents(se_SCT_norm_subset))
```

## LogNormalized data

Get the count matrix for LogNormalized data

```{r}
RPM_normalized <- LayerData(se_RPM_norm, assay = "RNA", layer = "counts")
head(RPM_normalized)
all(rownames(metadata_se) == colnames(RPM_normalized))  #see if cell barcodes match
```

Use sampled_cells to subset 

```{r}
RPM_normalized_subset <- RPM_normalized[, colnames(RPM_normalized) %in% sampled_cells]
```

```{r}
# Create an assay object before creating a Seurat object
RPM_normalized_assay <- CreateAssayObject(RPM_normalized_subset) 
se_RPM_norm_subset <- CreateSeuratObject(RPM_normalized_assay, meta.data = metadata)

# Check if the subsetting worked
Idents(se_RPM_norm_subset) <- se_RPM_norm_subset$Annotation
table(Idents(se_RPM_norm_subset))
```

# 5) PCA 

## For LogNormalized data

```{r}
se_RPM_norm_subset <- FindVariableFeatures(se_RPM_norm_subset) 
se_RPM_norm_subset <- ScaleData(se_RPM_norm_subset)
se_RPM_norm_subset <- RunPCA(se_RPM_norm_subset, verbose = FALSE)
```

## PCA on the SCTransform normalized dataset

```{r}
se_SCT_norm_subset <- FindVariableFeatures(se_SCT_norm_subset) 
se_SCT_norm_subset <- ScaleData(se_SCT_norm_subset)
se_SCT_norm_subset <- RunPCA(se_SCT_norm_subset, verbose = FALSE)
```

## Visualizing the PCA

## LogNormalized data

```{r}
# Get the PCA data
pca_data_RPM <- Embeddings(se_RPM_norm_subset, reduction = "pca")
pca_df_RPM <- as.data.frame(pca_data_RPM)
pca_df_RPM$barcodes <- rownames(pca_df_RPM) #rows not cols, switched during PCA computation

# Add metadata (from se to preserve consistency)
metadata_pca_RPM <- se_RPM_norm_subset@meta.data
pca_df_RPM <- cbind(pca_df_RPM, metadata_pca_RPM)
```

Colored by sequencing depth:

```{r}
ggplot(pca_df_RPM, aes(x = PC_1, y = PC_2, color = nCount_RNA)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Sequencing Depth", title = "RPM Normalized Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_RPM.png",
  plot = last_plot()
  )
```

Colored by cell type:

```{r}
ggplot(pca_df_RPM, aes(x = PC_1, y = PC_2, color = Annotation)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Cell Type", title = "RPM Normalized Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_RPM_annot.png",
  plot = last_plot()
  )
```

## SCTransformed data

```{r}
# Get the PCA data
pca_data_SCT <- Embeddings(se_SCT_norm_subset, reduction = "pca")
pca_df_SCT <- as.data.frame(pca_data_SCT)
pca_df_SCT$barcodes <- rownames(pca_df_SCT) 

# Add metadata (from se to preserve consistency)
metadata_pca_SCT <- se_SCT_norm_subset@meta.data
pca_df_SCT <- cbind(pca_df_SCT, metadata_pca_SCT)
```

Colored by sequencing depth:

```{r}
ggplot(pca_df_SCT, aes(x = PC_1, y = PC_2, color = nCount_RNA)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Sequencing Depth", title = "SCTransformed Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_SCT.png",
  plot = last_plot()
  )
```

Colored by cell type:

```{r}
ggplot(pca_df_SCT, aes(x = PC_1, y = PC_2, color = Annotation)) +
  geom_point() +
  theme_minimal() +
  labs(x = "PC1", y = "PC2", color = "Cell Type", title = "SCTransformed Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "PCA_SCT_annot.png",
  plot = last_plot()
  )
```

## Elbow plot to visualise SDs of PCs - helps us understand how many PCs we need to capture the most variance in the data

## SCtransformed

```{r}
ElbowPlot(se_SCT_norm_subset)
ggsave(
  "elbowplot_SCT.png",
  plot = last_plot()
  )
```

## Quantitative interpretation - makes it easier to choose the appropriate amount of PCs we need
https://hbctraining.github.io/scRNA-seq/lessons/elbow_plot_metric.html

1) The point where the principal components only contribute 5% of standard deviation and the principal components cumulatively contribute 90% of the standard deviation.

2) The point where the percent change in variation between the consecutive PCs is less than 0.1%. 

```{r}

# Determine percent of variation associated with each PC
pct_SCT <- se_SCT_norm_subset[["pca"]]@stdev / sum(se_SCT_norm_subset[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu_SCT <- cumsum(pct_SCT)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
PC1_SCT <- which(cumu_SCT > 90 & pct_SCT < 5)[1]

PC1_SCT

# Determine the difference between variation of PC and subsequent PC
PC_diff_SCT <- sort(which((pct_SCT[1:length(pct_SCT) - 1] - pct_SCT[2:length(pct_SCT)]) > 0.1), decreasing = T)[1] + 1

# Last point where change of % of variation is more than 0.1%.
PC_diff_SCT
```

## LogNormalized


```{r}
ElbowPlot(se_RPM_norm_subset)
ggsave(
  "elbowplot_RPM.png",
  plot = last_plot()
  )
```

```{r}

# Determine percent of variation associated with each PC
pct_RPM <- se_RPM_norm_subset[["pca"]]@stdev / sum(se_RPM_norm_subset[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu_RPM <- cumsum(pct_RPM)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
PC1_RPM <- which(cumu_RPM > 90 & pct_RPM < 5)[1]

PC1_RPM

# Determine the difference between variation of PC and subsequent PC
PC_diff_RPM <- sort(which((pct_RPM[1:length(pct_RPM) - 1] - pct_RPM[2:length(pct_RPM)]) > 0.1), decreasing = T)[1] + 1

# Last point where change of % of variation is more than 0.1%.
PC_diff_RPM
```

# 6) Run a UMAP on the PCA data - set a threshold for the PCs, e.g. 90% (based on the elbow plot)

Color UMAP by cell type - write a fn to quantify it (to see the separation of clusters) 

## SCTranformed

```{r}
se_SCT_norm_subset <- RunUMAP(se_SCT_norm_subset, dims = 1:15)  # Adjust based on elbow plot
```

```{r}
umap_df_SCT <- as.data.frame(Embeddings(se_SCT_norm_subset, reduction = "umap"))
```

```{r}
# Add metadata (from se to preserve consistency)
metadata_umap_SCT <- se_SCT_norm_subset@meta.data
umap_df_SCT <- cbind(umap_df_SCT, metadata_umap_SCT)
```

# Visualize UMAP

## Colored by sequencing depth

```{r}
ggplot(umap_df_SCT, aes(x = umap_1, y = umap_2, color = nCount_RNA)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Sequencing Depth", title = "SCTransformed Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "umap_SCT_seqdepth.png",
  plot = last_plot()
  )
```

## Colored by cell type

```{r}
ggplot(umap_df_SCT, aes(x = umap_1, y = umap_2, color = Annotation)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Cell Type", title = "SCTransformed Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5),
    legend.key.size = unit(0.5, "cm")
    )
ggsave(
  "umap_SCT_ctype.png",
  plot = last_plot()
  )
```

## LogNormalized

```{r}
se_RPM_norm_subset <- RunUMAP(se_RPM_norm_subset, dims = 1:19)  # Adjust based on elbow plot
```

```{r}
umap_df_RPM <- as.data.frame(Embeddings(se_RPM_norm_subset, reduction = "umap"))
```

```{r}
# Add metadata (from se to preserve consistency)
metadata_umap_RPM <- se_RPM_norm_subset@meta.data
umap_df_RPM <- cbind(umap_df_RPM, metadata_umap_RPM)
```

## Visualize UMAP

## Colored by sequencing depth

```{r}
ggplot(umap_df_RPM, aes(x = umap_1, y = umap_2, color = nCount_RNA)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Sequencing Depth", title = "RPM Normalized Data") +
  scale_color_viridis_c() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5))
ggsave(
  "umap_RPM_seqdepth.png",
  plot = last_plot()
  )
```

## Colored by cell type

```{r}
ggplot(umap_df_RPM, aes(x = umap_1, y = umap_2, color = Annotation)) +
  geom_point() +
  theme_minimal() +
  labs(x = "UMAP1", y = "UMAP2", color = "Cell Type", title = "RPM Normalized Data") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black", size = 0.5),
    legend.key.size = unit(0.5, "cm")
    )
ggsave(
  "umap_RPM_ctype.png",
  plot = last_plot()
  )
```

# 7) Quantifying the UMAP - separation by clusters

*Using centroids - not very informative

First, need to identify the centroids of each cluster. Then, use these to calculate the distances between the clusters.

SCTransformed

```{r}
# Find the centroids for each cluster
cluster_centroids_SCT <- aggregate(umap_df_SCT, by = list(Cluster = Idents(se_norm_subset)), FUN = mean)
cluster_names_SCT <- cluster_centroids_SCT[,1]

# Now find the distances between centroids of the clusters
quant_umap_SCT <- as.matrix(dist(cluster_centroids_SCT[, -1]))
rownames(quant_umap_SCT) <- cluster_names_SCT
colnames(quant_umap_SCT) <- cluster_names_SCT

quant_umap_SCT
```


```{r}
pheatmap(quant_umap_SCT,
         color = viridis(100),
         main = "Cluster Distances in the UMAP Space",
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean")
```

LogNormalized

```{r}
# Find the centroids for each cluster
cluster_centroids_RPM <- aggregate(umap_df_RPM, by = list(Cluster = Idents(se_RPM_norm_subset)), FUN = mean)
cluster_names_RPM <- cluster_centroids_RPM[,1]

# Now find the distances between centroids of the clusters
quant_umap_RPM <- as.matrix(dist(cluster_centroids_RPM[, -1]))
rownames(quant_umap_RPM) <- cluster_names_RPM
colnames(quant_umap_RPM) <- cluster_names_RPM
```


```{r}
pheatmap(quant_umap_RPM,
         color = viridis(100),
         main = "Cluster Distances in the UMAP Space",
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean")
```

## UMAP quantificiation:
    - every cell in a cluster to other cells in the clusters
    - every cell to other cell type cells in other clusters
    - plot this to see the overlaps
 
```{r}
quant <- function(umap_df) {
  # Calculating the pairwise distances, storing in dist matrix
  dist_matrix <- as.matrix(dist(umap_df[, c("umap_1", "umap_2")]))
  
  # Calculating distances (to cells in same cluster and in other clusters) for a cell
  dist_calc <- function(i) {
    cell_cluster <- umap_df$Annotation[i]
    same_cluster <- umap_df$Annotation == cell_cluster #logical vector
    
    # Storing two vectors of distances per cell in a list: 1) dist to cells in same cluster, 
    # 2) dist to cells in other clusters
    list(
      in_cluster = dist_matrix[i, same_cluster & (seq_along(same_cluster) != i)],
      other_clusters = dist_matrix[i, !same_cluster]
    )
  }
  
  # Apply the function to all cells
  results <- lapply(seq_len(nrow(umap_df)), dist_calc)
  names(results) <- rownames(umap_df) #assign correct barcodes to the resulting list
  
  return(results)
}
```

## SCTransformed

```{r}
# Apply the quantification fn 
quant_SCT <- quant(umap_df_SCT)
```

## Log normalized

```{r}
# Apply the quantification fn 
quant_RPM <- quant(umap_df_RPM)
```


# kNN clustering — for every cell, 5 closest neighbors

## LogNormalized

```{r}
se_RPM_norm_subset <- FindNeighbors(
  se_RPM_norm_subset,
  reduction = "umap",
  dims = 1:2,
  k.param = 5)
```

Visualize:

```{r}
# Extract kNN data
knn_RPM <- LayerData(se_RPM_norm_subset, graph = "RNA_snn") 

# Identifying highly connected cells
RPM_cell_connections <- colSums(knn_RPM)
RPM_top_cells <- names(sort(RPM_cell_connections, decreasing = TRUE)[1:10])
```

## SCTransformed

```{r}
se_SCT_norm_subset <- FindNeighbors(
  se_SCT_norm_subset,
  reduction = "umap",
  dims = 1:2,
  k.param = 5)
```


```{r}
# Extract kNN data
knn_SCT <- LayerData(se_SCT_norm_subset, graph = "RNA_snn") 

# Identifying highly connected cells
SCT_cell_connections <- colSums(knn_SCT)
SCT_top_cells <- names(sort(SCT_cell_connections, decreasing = TRUE)[1:10])
```

# Data file contains:

## se and se_subset

- original and filtered Seurat objects

## count_matrix_filt

- filtered count data extracted from se_subset

## SCT_normalized, RPM_normalized

- normalized count matrices

## se_RPM_norm_subset, se_SCT_norm_subset 

- QC done (filtered cells, genes, cells w/ high mito gene count) 
- kNN graphs
- RPM - normalization with Seurat NormalizeData (LogNormalize)
- SCT - SCTransformed (outside the Seurat object)
- UMAP and PCA dim reductions

## pca_df_RPM, pca_df_SCT, umap_df_RPM, umap_df_SCT

- extracted PCA and UMAP data


# Matrices exported for the projection step

## Log normalized and SCTransformed matrices

```{r}
saveRDS(RPM_normalized_subset, file = "RPM_normalized_matrix.Rdata")
saveRDS(SCT_normalized_subset, file = "SCT_normalized_matrix.Rdata")
```


## PCA embeddings







