---
title: "ISS Data Analysis and scRNA-seq Projections"
output: html_notebook
---

The purpose of this notebook is to process the ISS data, run PCA and UMAP for dimensionality reduction and visualization, import previously normalized scRNA-seq datasets and project these onto the ISS UMAP space. 

Quantile normalization is applied later - for comparison of results with and without data transformation, run the whole notebook.

## Load libraries

```{r}
library(Seurat)
library(Matrix)
library(ggplot2)
library(tidyverse)
library(dplyr)
library(gprofiler2)
library(sctransform)
library(magrittr)
library(umap)
library(stats)
library(pals)
library(preprocessCore)
```

Create color palettes for plotting

```{r}
colors <- pals::glasbey(20)
sequential <- pals::cividis(100)
```


### Load the data

Data is available here: https://drive.google.com/drive/folders/1KfcSJUEDn3UJxFnsixnLXUjWTvWy3Ymm?usp=drive_link

Metadata: https://drive.google.com/drive/folders/1XHkZ_wsk9y_LQwP4ydhZ5cj7kCtyat9v?usp=drive_link

Change the paths accordingly:

```{r}
ISS_matrix <- readMM("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/matrix.mtx")
```

```{r}
features <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/features.tsv", sep="\t", header=FALSE, stringsAsFactors=FALSE)

barcodes <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/barcodes.tsv", header=FALSE, stringsAsFactors=FALSE)

metadata <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/Fig3e-j_Xenium.csv", stringsAsFactors=FALSE, sep=";", header=TRUE, row.names=1)
```

Assign barcodes and genes as names for the count matrix.

```{r}
rownames(ISS_matrix) <- features$V2  # gene symbols as row names
colnames(ISS_matrix) <- barcodes$V1  # barcodes as column names
colnames(features) <- c("ensembl","gene","gene_expression")
```


### Load normalized scRNA-seq count matrices

Either use the files obtained from the scRNA-seq analysis or refer to the following Google drive folder: https://drive.google.com/drive/folders/1UPB5Fgct1HTnDXq8rl1scaiNogfGN5aB?usp=drive_link


```{r}
RPM_norm_matrix_sc <- readRDS(file = "C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/RPM_norm_matrix_sc.rds")
SCT_norm_matrix_sc <- readRDS(file = "C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/ISS/data/SCT_norm_matrix_sc.rds")
```

```{r}
# Load scRNA-seq metadata
metadata_sc <- read.table("C:/Users/liisa/OneDrive/Töölaud/Project-in-MLS/scRNA-seq/data/Fig2a_scFFPE-seq_UMAP.csv", stringsAsFactors=FALSE, sep=";", header=TRUE, row.names=1)
```

## 1) QC 

### Filter both the matrix and the feature data frame to exclude probes

First, filter the features data frame.

```{r}
features <- filter(features, gene_expression == "Gene Expression")
```

Filter the genes in the count matrix, based on previous filtering.

```{r}
genes_list <- features$gene

ISS_matrix_filtered <- ISS_matrix[genes_list, ]
```

Create a separate column for cell barcodes to keep track of them; remove unlabeled cells.

```{r}
metadata$barcodes <- rownames(metadata)
metadata <- metadata[metadata$Cluster != "Unlabeled", ]
```

Filter and keep cells in metadata and matrix that are present in both.

```{r}
same_barcodes <- intersect(metadata$barcodes, colnames(ISS_matrix_filtered))
ISS_matrix_filtered <- ISS_matrix_filtered[, same_barcodes]
metadata <- metadata[metadata$barcodes %in% same_barcodes, ]
```

### Filter genes

- Find common genes with the scRNA-seq data: use one of the normalized matrices for filtering since they both contain the same genes.

- Apply the filtering on all 3 matrices.

```{r}
common_genes <- intersect(rownames(ISS_matrix_filtered), rownames(RPM_norm_matrix_sc))
ISS_matrix_filtered <- ISS_matrix_filtered[common_genes, ]
RPM_norm_matrix_sc <- RPM_norm_matrix_sc[common_genes,]
SCT_norm_matrix_sc <- SCT_norm_matrix_sc[common_genes,]
```


### Subset the data to only include 100 cells from each cell type


```{r}
sampled_cells <- metadata %>%
  group_by(Cluster) %>%
  slice_sample(n = 100, replace = FALSE) %>% 
  pull(barcodes)
```

```{r}
ISS_matrix_filtered <- ISS_matrix_filtered[, colnames(ISS_matrix_filtered) %in% sampled_cells]
table(metadata$Cluster[metadata$barcodes %in% colnames(ISS_matrix_filtered)])
```

Subset metadata based on the filtered count matrix

```{r}
barcodes <- intersect(colnames(ISS_matrix_filtered), metadata$barcodes)
metadata_subset <- metadata[barcodes, ]
```


# 2) Normalization 

## Method 1: log2(x+1) normalization - not RPM cause ISS is targeted

```{r}
LOG_ISS_matrix <- log2(ISS_matrix_filtered+1)
```

## Method 2: SCTransform

```{r}
ISS_matrix_filtered <- as(ISS_matrix_filtered, "CsparseMatrix")
```

```{r}
SCT_ISS_matrix <- sctransform::vst(ISS_matrix_filtered)$y
```

SCTransform removed one gene, filter genes for all matrices again

```{r}
common_genes_2 <- intersect(rownames(SCT_ISS_matrix), rownames(RPM_norm_matrix_sc))
SCT_ISS_matrix <- SCT_ISS_matrix[common_genes_2, ]
LOG_ISS_matrix <- LOG_ISS_matrix[common_genes_2, ]
RPM_norm_matrix_sc <- RPM_norm_matrix_sc[common_genes_2, ]
SCT_norm_matrix_sc <- SCT_norm_matrix_sc[common_genes_2, ]
```


# 3) PCA using prcomp

```{r}
t_LOG_ISS_matrix <- t(LOG_ISS_matrix)
t_SCT_ISS_matrix <- t(SCT_ISS_matrix)
```


## Log transformed

```{r}
LOG_PCA <- prcomp(t_LOG_ISS_matrix)
```

## SCTransformed

```{r}
SCT_PCA <- prcomp(t_SCT_ISS_matrix)
```


# 4) Visualize PCA

## Log transformed

```{r}
# Extract PC scores from prcomp object
LOG_PCA_scores <- as.data.frame(LOG_PCA$x)
```

```{r}
LOG_PCA_scores <- LOG_PCA_scores[,-51:-295] # we don't need all of the PCs for further analysis

# Reorder
LOG_metadata_subset <- metadata_subset[match(rownames(LOG_PCA_scores), metadata_subset$barcodes), ]

# Add metadata to the PC scores
LOG_PCA_scores$Cluster <- LOG_metadata_subset$Cluster
LOG_PCA_scores$transcript_counts <- LOG_metadata_subset$transcript_counts
```


```{r}
LOG_PCA_scores$Cluster <- as.factor(LOG_PCA_scores$Cluster) # convert to factor to facilitate plotting
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = LOG_PCA_scores$PC1,
  y = LOG_PCA_scores$PC2,
  col = colors[LOG_PCA_scores$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "PC1",
  ylab = "PC2",
)

coord <- par("usr") # get plot coordinates
legend(x = coord[2], y = coord[4],
       legend = levels(LOG_PCA_scores$Cluster),
       col = colors[1:nlevels(LOG_PCA_scores$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # smaller text
       pt.cex = 1,  # also smaller point size
       xpd = TRUE) # legend outside of the plot bounds
```
```{r}
# Rescale to fit the color palette because there are more values than colors 
LOG_rescaled_counts <- scales::rescale(LOG_PCA_scores$transcript_counts, to = c(1, 100))
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = LOG_PCA_scores$PC1,
  y = LOG_PCA_scores$PC2,
  col = sequential[LOG_rescaled_counts],
  pch = 19,
  cex = 0.5,
  xlab = "PC1",
  ylab = "PC2",
)

coord <- par("usr") # get plot coordinates
legend_colors <- sequential[seq(1, 100, length.out = 5)]
legend_values <- seq(min(LOG_PCA_scores$transcript_counts), 
                           max(LOG_PCA_scores$transcript_counts), 
                           length.out = 5)


legend(x = coord[2]+1, y = coord[4],
       legend = legend_values,
       fill = legend_colors,
       title = "Sequencing Depth",
       bty = "n",
       cex = 0.8, 
       xpd = TRUE) # legend outside of the plot bounds
```

## SCTransformed

```{r}
# Extract PC scores from prcomp object
SCT_PCA_scores <- as.data.frame(SCT_PCA$x) 
SCT_PCA_scores <- SCT_PCA_scores[,-51:-295]
```

```{r}
# Check if the order of cells in metadata matches the order in PCA df
SCT_metadata_subset <- metadata_subset[match(rownames(SCT_PCA_scores), metadata_subset$barcodes), ]

# Add metadata to the PC scores
SCT_PCA_scores$Cluster <- SCT_metadata_subset$Cluster
SCT_PCA_scores$transcript_counts <- SCT_metadata_subset$transcript_counts
```


```{r}
SCT_PCA_scores$Cluster <- as.factor(SCT_PCA_scores$Cluster)
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = SCT_PCA_scores$PC1,
  y = SCT_PCA_scores$PC2,
  col = colors[SCT_PCA_scores$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "PC1",
  ylab = "PC2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(SCT_PCA_scores$Cluster),
       col = colors[1:nlevels(SCT_PCA_scores$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```

```{r}
# Rescale to fit the color palette because there are more values than colors 
SCT_rescaled_counts <- scales::rescale(SCT_PCA_scores$transcript_counts, to = c(1, 100))
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = SCT_PCA_scores$PC1,
  y = SCT_PCA_scores$PC2,
  col = sequential[SCT_rescaled_counts],
  pch = 19,
  cex = 0.5,
  xlab = "PC1",
  ylab = "PC2",
)

coord <- par("usr") # get plot coordinates
legend_colors <- sequential[seq(1, 100, length.out = 5)]
legend_values <- seq(min(SCT_PCA_scores$transcript_counts), 
                           max(SCT_PCA_scores$transcript_counts), 
                           length.out = 5)


legend(x = coord[2]+1, y = coord[4],
       legend = legend_values,
       fill = legend_colors,
       title = "Sequencing Depth",
       bty = "n",
       cex = 0.8, 
       xpd = TRUE) # legend outside of the plot bounds
```

## Elbowplots

```{r}
LOG_PCA_scores_e <- as.matrix(LOG_PCA_scores[,-51:-52])

# Also extract stdev from prcomp object
LOG_PCA_stdev <- LOG_PCA$sdev

LOG_ISS_assay <- CreateAssayObject(LOG_ISS_matrix)
LOG_ISS_se <- CreateSeuratObject(LOG_ISS_assay, 
                                 project = "XeniumProject",
                                 assay = "Xenium",
                                 meta.data = metadata)
LOG_ISS_se[["pca"]] <- CreateDimReducObject(embeddings = LOG_PCA_scores_e, key = "PC_", assay = DefaultAssay(LOG_ISS_se)) # add PCA
LOG_ISS_se[["pca"]]@stdev <- LOG_PCA_stdev #add stdev
```

```{r}
ElbowPlot(LOG_ISS_se) 
```

LOG: PC = 15


```{r}
SCT_PCA_scores_e <- as.matrix(SCT_PCA_scores[,-51:-52])

# Also extract stdev from prcomp object
SCT_PCA_stdev <- SCT_PCA$sdev

SCT_ISS_assay <- CreateAssayObject(SCT_ISS_matrix)
SCT_ISS_se <- CreateSeuratObject(SCT_ISS_assay, 
                                 project = "XeniumProject",
                                 assay = "Xenium",
                                 meta.data = metadata)
SCT_ISS_se[["pca"]] <- CreateDimReducObject(embeddings = SCT_PCA_scores_e, key = "PC_", assay = DefaultAssay(SCT_ISS_se)) # add PCA
SCT_ISS_se[["pca"]]@stdev <- SCT_PCA_stdev #add stdev
```

```{r}
ElbowPlot(SCT_ISS_se) 
```

SCT: PC = 16

# 5) Run UMAP


```{r}
LOG_PCA_scores_e <- LOG_PCA_scores_e[,1:15]
SCT_PCA_scores_e <- SCT_PCA_scores_e[,1:16]

LOG_UMAP <- umap(LOG_PCA_scores_e)
SCT_UMAP <- umap(SCT_PCA_scores_e)
```


# Visualize UMAPs

## SCTransformed

```{r}
# Extract umap coordinates for plotting
SCT_UMAP_df <- as.data.frame(SCT_UMAP$layout) 
```

```{r}
# Reorder metadata to match
SCT_metadata_subset <- SCT_metadata_subset[match(rownames(SCT_UMAP_df), SCT_metadata_subset$barcodes), ]
```

```{r}
# Add metadata to the PC scores
SCT_UMAP_df$Cluster <- SCT_metadata_subset$Cluster
SCT_UMAP_df$transcript_counts <- SCT_metadata_subset$transcript_counts
colnames(SCT_UMAP_df)[1] <- "UMAP1"  
colnames(SCT_UMAP_df)[2] <- "UMAP2"
```


```{r}
SCT_UMAP_df$Cluster <- as.factor(SCT_UMAP_df$Cluster)
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = SCT_UMAP_df$UMAP1,
  y = SCT_UMAP_df$UMAP2,
  col = colors[SCT_UMAP_df$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(SCT_UMAP_df$Cluster),
       col = colors[1:nlevels(SCT_UMAP_df$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```
```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = SCT_UMAP_df$UMAP1,
  y = SCT_UMAP_df$UMAP2,
  col = sequential[SCT_rescaled_counts],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates
legend_colors <- sequential[seq(1, 100, length.out = 5)]
legend_values <- seq(min(SCT_UMAP_df$transcript_counts), 
                           max(SCT_UMAP_df$transcript_counts), 
                           length.out = 5)


legend(x = coord[2]+1, y = coord[4],
       legend = legend_values,
       fill = legend_colors,
       title = "Sequencing Depth",
       bty = "n",
       cex = 0.8, 
       xpd = TRUE) # legend outside of the plot bounds
```

## Log Normalized

```{r}
# Extract umap coordinates for plotting
LOG_UMAP_df <- as.data.frame(LOG_UMAP$layout) 
```

```{r}
# Add metadata to the PC scores
LOG_metadata_subset <- LOG_metadata_subset[match(rownames(LOG_UMAP_df), LOG_metadata_subset$barcodes), ]
LOG_UMAP_df$Cluster <- LOG_metadata_subset$Cluster
LOG_UMAP_df$transcript_counts <- LOG_metadata_subset$transcript_counts
colnames(LOG_UMAP_df)[1] <- "UMAP1"  
colnames(LOG_UMAP_df)[2] <- "UMAP2"
```


```{r}
LOG_UMAP_df$Cluster <- as.factor(LOG_UMAP_df$Cluster)
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = LOG_UMAP_df$UMAP1,
  y = LOG_UMAP_df$UMAP2,
  col = colors[LOG_UMAP_df$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(LOG_UMAP_df$Cluster),
       col = colors[1:nlevels(LOG_UMAP_df$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = LOG_UMAP_df$UMAP1,
  y = LOG_UMAP_df$UMAP2,
  col = sequential[LOG_rescaled_counts],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates
legend_colors <- sequential[seq(1, 100, length.out = 5)]
legend_values <- seq(min(LOG_UMAP_df$transcript_counts), 
                           max(LOG_UMAP_df$transcript_counts), 
                           length.out = 5)


legend(x = coord[2]+1, y = coord[4],
       legend = legend_values,
       fill = legend_colors,
       title = "Sequencing Depth",
       bty = "n",
       cex = 0.8, 
       xpd = TRUE) # legend outside of the plot bounds
```

# Projecting normalized scRNA-seq data on to the PCA from the ISS data

# 1) normalized scRNA-seq matrix onto the PCA from ISS 

## Log normalized

Since we need to project scRNA-seq data onto the same PCA space, rather than the coordinates associated with the cells, we need to know which genes and how contribute to the PCA components (therefore use rotation).

```{r}
LOG_PCA_rotation <- as.matrix(LOG_PCA$rotation)
```

Project the data: 

https://stats.stackexchange.com/questions/2592/how-to-project-a-new-vector-onto-pca-space

Transpose scRNA-seq matrix so dimensions work when multiplying with the PCA coordinates from ISS data

```{r}
RPM_sc_projected <- t(RPM_norm_matrix_sc) %*% LOG_PCA_rotation
```


## SCTransformed

```{r}
SCT_PCA_rotation <- as.matrix(SCT_PCA$rotation)
```

```{r}
SCT_sc_projected <- t(SCT_norm_matrix_sc) %*% SCT_PCA_rotation
```


# Project scRNA-seq data in the ISS PCA space onto the ISS UMAP space

## SCTransformed

```{r}
SCT_sc_projected <- SCT_sc_projected[,-17:-295] # use as many PCs as for ISS UMAP
SCT_sc_UMAP_projected <- predict(SCT_UMAP, SCT_sc_projected)
colnames(SCT_sc_UMAP_projected) <- c("UMAP1", "UMAP2")
```

```{r}
# Reorder metadata and readd annotations
SCT_metadata_sc <- metadata_sc[rownames(metadata_sc) %in% rownames(SCT_sc_UMAP_projected), ]
SCT_sc_UMAP_projected <- cbind(SCT_sc_UMAP_projected, Annotation = SCT_metadata_sc$Annotation)
SCT_sc_UMAP_projected_df <- as.data.frame(SCT_sc_UMAP_projected)
```


```{r}
SCT_sc_UMAP_projected_df$Annotation <- as.factor(SCT_sc_UMAP_projected_df$Annotation)
```


```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = SCT_sc_UMAP_projected_df$UMAP1,
  y = SCT_sc_UMAP_projected_df$UMAP2,
  col = colors[SCT_sc_UMAP_projected_df$Annotation],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(SCT_sc_UMAP_projected_df$Annotation),
       col = colors[1:nlevels(SCT_sc_UMAP_projected_df$Annotation)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```

We can see that without using quantile normalization for transforming ISS and scRNA-seq datasets, the integrated data are not interpretable. 


## Log Normalized

```{r}
RPM_sc_projected <- RPM_sc_projected[,-16:-295]
RPM_sc_UMAP_projected <- predict(LOG_UMAP, RPM_sc_projected)
colnames(RPM_sc_UMAP_projected) <- c("UMAP1", "UMAP2")
```

```{r}
# Reorder metadata and readd annotations
RPM_metadata_sc <- metadata_sc[rownames(metadata_sc) %in% rownames(RPM_sc_UMAP_projected), ]
RPM_sc_UMAP_projected <- cbind(RPM_sc_UMAP_projected, Annotation = RPM_metadata_sc$Annotation)
RPM_sc_UMAP_projected_df <- as.data.frame(RPM_sc_UMAP_projected)
```

Plot projected UMAP:

```{r}
RPM_sc_UMAP_projected_df$Annotation <- as.factor(RPM_sc_UMAP_projected_df$Annotation)
```


```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = RPM_sc_UMAP_projected_df$UMAP1,
  y = RPM_sc_UMAP_projected_df$UMAP2,
  col = colors[RPM_sc_UMAP_projected_df$Annotation],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(RPM_sc_UMAP_projected_df$Annotation),
       col = colors[1:nlevels(RPM_sc_UMAP_projected_df$Annotation)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```


## Save datasets for further analysis steps

```{r}
LOG_ISS_matrix_sparse <- as(LOG_ISS_matrix,"CsparseMatrix")
SCT_ISS_matrix_sparse <- as(SCT_ISS_matrix,"CsparseMatrix")
```


```{r}
saveRDS(LOG_ISS_matrix_sparse, file = "LOG_ISS_matrix.rds")
saveRDS(SCT_ISS_matrix_sparse, file = "SCT_ISS_matrix.rds")
```


### Apply quantile normalization

Quantile normalization:

1) Concatenate datasets

```{r}
combined_RPM_LOG <- cbind(RPM_norm_matrix_sc, LOG_ISS_matrix)
```

2) Normalize

https://www.rdocumentation.org/packages/preprocessCore/versions/1.34.0/topics/normalize.quantiles

```{r}
qnormalized_RPM_LOG <- normalize.quantiles(as.matrix(combined_RPM_LOG))
```

3) Separate the datasets again

```{r}
q_RPM_norm_matrix_sc <- qnormalized_RPM_LOG[, 1:ncol(RPM_norm_matrix_sc)]
q_LOG_ISS_matrix <- qnormalized_RPM_LOG[, (ncol(RPM_norm_matrix_sc) + 1):ncol(combined_RPM_LOG)]
```

Assign gene names and cell barcodes again

```{r} 
rownames(q_LOG_ISS_matrix) <- common_genes_2
colnames(q_LOG_ISS_matrix) <- barcodes
```

Check that the barcodes match

```{r}
sc_barcodes <- colnames(RPM_norm_matrix_sc)
all(colnames(SCT_norm_matrix_sc) == sc_barcodes)
```

Repeat for scRNA-seq data

```{r} 
rownames(q_RPM_norm_matrix_sc) <- common_genes_2
colnames(q_RPM_norm_matrix_sc) <- sc_barcodes
```



### Repeat for SCT

```{r}
combined_SCT <- cbind(SCT_norm_matrix_sc, SCT_ISS_matrix)
```

Normalize

```{r}
qnormalized_SCT <- normalize.quantiles(as.matrix(combined_SCT))
```

Separate the datasets again

```{r}
q_SCT_norm_matrix_sc <- qnormalized_SCT[, 1:ncol(SCT_norm_matrix_sc)]
q_SCT_ISS_matrix <- qnormalized_SCT[, (ncol(SCT_norm_matrix_sc) + 1):ncol(combined_SCT)]
```

Assign gene names and cell barcodes again

```{r} 
rownames(q_SCT_ISS_matrix) <- common_genes_2
colnames(q_SCT_ISS_matrix) <- barcodes
```

```{r} 
rownames(q_SCT_norm_matrix_sc) <- common_genes_2
colnames(q_SCT_norm_matrix_sc) <- sc_barcodes
```


### PCA and UMAP on quantile normalized ISS data

```{r}
t_q_LOG_ISS_matrix <- t(q_LOG_ISS_matrix)
t_q_SCT_ISS_matrix <- t(q_SCT_ISS_matrix)
```


```{r}
q_LOG_PCA <- prcomp(t_q_LOG_ISS_matrix, rank = 30)
```

```{r}
q_SCT_PCA <- prcomp(t_q_SCT_ISS_matrix, rank = 30)
```


### Visualize PCA

Log transformed

```{r}
# Extract PC scores from prcomp object
q_LOG_PCA_scores <- as.data.frame(q_LOG_PCA$x)
```

```{r}
# Reorder
q_LOG_metadata_subset <- metadata_subset[match(rownames(q_LOG_PCA_scores), metadata_subset$barcodes), ]

# Add metadata to the PC scores
q_LOG_PCA_scores$Cluster <- q_LOG_metadata_subset$Cluster
q_LOG_PCA_scores$transcript_counts <- q_LOG_metadata_subset$transcript_counts
```


```{r}
q_LOG_PCA_scores$Cluster <- as.factor(q_LOG_PCA_scores$Cluster)
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = q_LOG_PCA_scores$PC1,
  y = q_LOG_PCA_scores$PC2,
  col = colors[q_LOG_PCA_scores$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "PC1",
  ylab = "PC2",
)

coord <- par("usr") # get plot coordinates
legend(x = coord[2], y = coord[4],
       legend = levels(q_LOG_PCA_scores$Cluster),
       col = colors[1:nlevels(q_LOG_PCA_scores$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # smaller text
       pt.cex = 1,  # also smaller point size
       xpd = TRUE) # legend outside of the plot bounds
```

SCT

```{r}
# Extract PC scores from prcomp object
q_SCT_PCA_scores <- as.data.frame(q_SCT_PCA$x) 
```

```{r}
# Check if the order of cells in metadata matches the order in PCA df
q_SCT_metadata_subset <- metadata_subset[match(rownames(q_SCT_PCA_scores), metadata_subset$barcodes), ]

# Add metadata to the PC scores
q_SCT_PCA_scores$Cluster <- q_SCT_metadata_subset$Cluster
q_SCT_PCA_scores$transcript_counts <- q_SCT_metadata_subset$transcript_counts
```


```{r}
q_SCT_PCA_scores$Cluster <- as.factor(q_SCT_PCA_scores$Cluster)
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = q_SCT_PCA_scores$PC1,
  y = q_SCT_PCA_scores$PC2,
  col = colors[q_SCT_PCA_scores$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "PC1",
  ylab = "PC2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(q_SCT_PCA_scores$Cluster),
       col = colors[1:nlevels(q_SCT_PCA_scores$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```

### Elbowplots

```{r}
q_LOG_PCA_scores_e <- as.matrix(q_LOG_PCA_scores[,-31:-32])

# Also extract stdev from prcomp object
q_LOG_PCA_stdev <- q_LOG_PCA$sdev

q_LOG_ISS_assay <- CreateAssayObject(q_LOG_ISS_matrix)
q_LOG_ISS_se <- CreateSeuratObject(q_LOG_ISS_assay, 
                                 project = "XeniumProject",
                                 assay = "Xenium",
                                 meta.data = metadata)
q_LOG_ISS_se[["pca"]] <- CreateDimReducObject(embeddings = q_LOG_PCA_scores_e, key = "PC_", assay = DefaultAssay(q_LOG_ISS_se)) # add PCA
q_LOG_ISS_se[["pca"]]@stdev <- q_LOG_PCA_stdev #add stdev
```

```{r}
ElbowPlot(q_LOG_ISS_se) 
```

LOG PC = 15


```{r}
q_SCT_PCA_scores_e <- as.matrix(q_SCT_PCA_scores[,-31:-32])

# Also extract stdev from prcomp object
q_SCT_PCA_stdev <- q_SCT_PCA$sdev

q_SCT_ISS_assay <- CreateAssayObject(q_SCT_ISS_matrix)
q_SCT_ISS_se <- CreateSeuratObject(q_SCT_ISS_assay, 
                                 project = "XeniumProject",
                                 assay = "Xenium",
                                 meta.data = metadata)
q_SCT_ISS_se[["pca"]] <- CreateDimReducObject(embeddings = q_SCT_PCA_scores_e, key = "PC_", assay = DefaultAssay(q_SCT_ISS_se)) # add PCA
q_SCT_ISS_se[["pca"]]@stdev <- q_SCT_PCA_stdev #add stdev
```

```{r}
ElbowPlot(q_SCT_ISS_se) 
```

SCT PC = 17

### Run UMAP

```{r}
q_LOG_PCA_scores_e <- q_LOG_PCA_scores_e[,1:15]
q_SCT_PCA_scores_e <- q_SCT_PCA_scores_e[,1:17]

q_LOG_UMAP <- umap(q_LOG_PCA_scores_e)
q_SCT_UMAP <- umap(q_SCT_PCA_scores_e)
```

LOG UMAP:

```{r}
# Extract umap coordinates for plotting
q_LOG_UMAP_df <- as.data.frame(q_LOG_UMAP$layout) 
```

```{r}
# Reorder to match
q_LOG_metadata_subset <- q_LOG_metadata_subset[match(rownames(q_LOG_UMAP_df), q_LOG_metadata_subset$barcodes), ]
```

```{r}
# Add metadata to the PC scores
q_LOG_UMAP_df$Cluster <- q_LOG_metadata_subset$Cluster
q_LOG_UMAP_df$transcript_counts <- q_LOG_metadata_subset$transcript_counts
colnames(q_LOG_UMAP_df)[1] <- "UMAP1"  
colnames(q_LOG_UMAP_df)[2] <- "UMAP2"
```


```{r}
q_LOG_UMAP_df$Cluster <- as.factor(q_LOG_UMAP_df$Cluster)
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = q_LOG_UMAP_df$UMAP1,
  y = q_LOG_UMAP_df$UMAP2,
  col = colors[q_LOG_UMAP_df$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(q_LOG_UMAP_df$Cluster),
       col = colors[1:nlevels(q_LOG_UMAP_df$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```

SCT UMAP:


```{r}
# Extract umap coordinates for plotting
q_SCT_UMAP_df <- as.data.frame(q_SCT_UMAP$layout) 
```

```{r}
# Reorder to match
q_SCT_metadata_subset <- q_SCT_metadata_subset[match(rownames(q_SCT_UMAP_df), q_SCT_metadata_subset$barcodes), ]
```

```{r}
# Add metadata to the PC scores
q_SCT_UMAP_df$Cluster <- q_SCT_metadata_subset$Cluster
q_SCT_UMAP_df$transcript_counts <- q_SCT_metadata_subset$transcript_counts
colnames(q_SCT_UMAP_df)[1] <- "UMAP1"  
colnames(q_SCT_UMAP_df)[2] <- "UMAP2"
```


```{r}
q_SCT_UMAP_df$Cluster <- as.factor(q_SCT_UMAP_df$Cluster)
```

```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = q_SCT_UMAP_df$UMAP1,
  y = q_SCT_UMAP_df$UMAP2,
  col = colors[q_SCT_UMAP_df$Cluster],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(q_SCT_UMAP_df$Cluster),
       col = colors[1:nlevels(q_SCT_UMAP_df$Cluster)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```

### Project scRNA-seq data to the ISS PCA space

```{r}
q_LOG_PCA_rotation <- as.matrix(q_LOG_PCA$rotation)
```

```{r}
q_RPM_sc_projected <- t(q_RPM_norm_matrix_sc) %*% q_LOG_PCA_rotation
```


```{r}
q_SCT_PCA_rotation <- as.matrix(q_SCT_PCA$rotation)
```

```{r}
q_SCT_sc_projected <- t(q_SCT_norm_matrix_sc) %*% q_SCT_PCA_rotation
```


# Project scRNA-seq data in the ISS PCA space onto the ISS UMAP space

## SCTransformed

```{r}
q_SCT_sc_projected <- q_SCT_sc_projected[,-18:-30] # use as many PCs as for ISS UMAP
q_SCT_sc_UMAP_projected <- predict(q_SCT_UMAP, q_SCT_sc_projected)
colnames(q_SCT_sc_UMAP_projected) <- c("UMAP1", "UMAP2")
```

```{r}
# Check rownames match
SCT_metadata_sc <- SCT_metadata_sc[match(rownames(q_SCT_sc_UMAP_projected), rownames(SCT_metadata_sc)), ]
all(rownames(q_SCT_sc_UMAP_projected) == rownames(SCT_metadata_sc))
```

```{r}
# Add annotations and convert to a df
q_SCT_sc_UMAP_projected <- cbind(q_SCT_sc_UMAP_projected, Annotation = SCT_metadata_sc$Annotation)
q_SCT_sc_UMAP_projected_df <- as.data.frame(q_SCT_sc_UMAP_projected)
```


```{r}
q_SCT_sc_UMAP_projected_df$Annotation <- as.factor(q_SCT_sc_UMAP_projected_df$Annotation)
```


```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = q_SCT_sc_UMAP_projected_df$UMAP1,
  y = q_SCT_sc_UMAP_projected_df$UMAP2,
  col = colors[q_SCT_sc_UMAP_projected_df$Annotation],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(q_SCT_sc_UMAP_projected_df$Annotation),
       col = colors[1:nlevels(q_SCT_sc_UMAP_projected_df$Annotation)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```

Repeat for RPM normalized

```{r}
q_RPM_sc_projected <- q_RPM_sc_projected[,-16:-30] # use as many PCs as for ISS UMAP
q_RPM_sc_UMAP_projected <- predict(q_LOG_UMAP, q_RPM_sc_projected)
colnames(q_RPM_sc_UMAP_projected) <- c("UMAP1", "UMAP2")
```

```{r}
RPM_metadata_sc <- RPM_metadata_sc[match(rownames(q_RPM_sc_UMAP_projected), rownames(RPM_metadata_sc)), ]
all(rownames(q_RPM_sc_UMAP_projected) == rownames(RPM_metadata_sc))
```

```{r}
q_RPM_sc_UMAP_projected <- cbind(q_RPM_sc_UMAP_projected, Annotation = RPM_metadata_sc$Annotation)
q_RPM_sc_UMAP_projected_df <- as.data.frame(q_RPM_sc_UMAP_projected)
```


```{r}
q_RPM_sc_UMAP_projected_df$Annotation <- as.factor(q_RPM_sc_UMAP_projected_df$Annotation)
```


```{r}
par(mar = c(5, 4, 2, 10), xpd = TRUE) # set plot margins
par(plt = c(0.12, 0.7, 0.2, 0.9)) # also edit plot area

plot(
  x = q_RPM_sc_UMAP_projected_df$UMAP1,
  y = q_RPM_sc_UMAP_projected_df$UMAP2,
  col = colors[q_RPM_sc_UMAP_projected_df$Annotation],
  pch = 19,
  cex = 0.5,
  xlab = "UMAP1",
  ylab = "UMAP2",
)

coord <- par("usr") # get plot coordinates

legend(x = coord[2], y = coord[4],
       legend = levels(q_RPM_sc_UMAP_projected_df$Annotation),
       col = colors[1:nlevels(q_RPM_sc_UMAP_projected_df$Annotation)], # get nr of clusters to apply right amount of colors
       pch = 19,
       title = "Cell Type",
       cex = 0.8,  # Adjust text size if needed
       pt.cex = 1,  # Adjust point size in legend if needed
       xpd = TRUE) # legend outside of the plot bounds
```



















